pkg/cluster/actions.go

```diff
@@ -0,0 +1,231 @@
+	actionImpls.Lock()
+	actionBuilderFunc, ok := actionImpls.impls[name]
+	actionImpls.Unlock()
+	if !ok {
+		return nil, fmt.Errorf("no Action implementation with name: %s", name)
+	}
+	return actionBuilderFunc(), nil
+}
+
+// NewExecutionPlan creates an execution plan by applying logical step/task
+// defined for each action to the actual cluster topology. As a result task
+// could be executed zero, one or more times according with the target nodes
+// selector defined for each task.
+// The execution plan is ordered, providing a predictable, "kubeadm friendly"
+// and consistent execution order; with this regard please note that the order
+// of actions is important, and it will be respected by planning.
+// TODO(fabrizio pandini): probably it will be necessary to add another criteria
+//     for ordering planned task for the most complex workflows (e.g.
+//     init-join-upgrade and then join again)
+//     e.g. it should be something like "action group" where each action
+//	   group is a list of actions
+func NewExecutionPlan(cfg *config.Config, actionNames []string) (ExecutionPlan, error) {
+	// for each actionName
+	var plan = ExecutionPlan{}
+	for i, name := range actionNames {
+		// get the action implementation instance
+		actionImpl, err := GetAction(name)
+		if err != nil {
+			return nil, err
+		}
+		// for each logical tasks defined for the action
+		for j, t := range actionImpl.Tasks() {
+			// get the list of target nodes in the current topology
+			targetNodes := t.TargetNodes(cfg)
+			for _, n := range targetNodes {
+				// creates the planned task
+				taskContext := &PlannedTask{
+					Node:        n,
+					Task:        t,
+					actionIndex: i,
+					taskIndex:   j,
+				}
+				plan = append(plan, taskContext)
+			}
+		}
+	}
+
+	// sorts the list of planned task ensuring a predictable, "kubeadm friendly"
+	// and consistent execution order
+	sort.Sort(plan)
+	return plan, nil
+}
+
+// Len of the ExecutionPlan.
+// It is required for making ExecutionPlan sortable.
+func (t ExecutionPlan) Len() int {
+	return len(t)
+}
+
+// Less return the lower between two elements of the ExecutionPlan, where the
+// lower element should be executed before the other.
+// It is required for making ExecutionPlan sortable.
+func (t ExecutionPlan) Less(i, j int) bool {
+	return t[i].ExecutionOrder() < t[j].ExecutionOrder()
+}
+
+// Swap two elements of the ExecutionPlan.
+// It is required for making ExecutionPlan sortable.
+func (t ExecutionPlan) Swap(i, j int) {
+	t[i], t[j] = t[j], t[i]
+}
+
+// ExecutionOrder returns a string that can be used for sorting planned tasks
+// into a predictable, "kubeadm friendly" and consistent order.
+// NB. we are using a string to combine all the item considered into something
+// that can be easily sorted using a lexicographical order
+func (p *PlannedTask) ExecutionOrder() string {
+	return fmt.Sprintf("Node.ProvisioningOrder: %d - Node.Name: %s - actionIndex: %d - taskIndex: %d",
```

Comment:
i might need a diagram to better understand how this work j/k.
in all seriousness, i think the sorting introduces a non-deterministic factor that could be avoided.

some questions:
- why do we even need the sorting?
- what if we want to execute tasks with an order not based on the node names?

Reply:
> why do we even need the sorting?
> what if we want to execute tasks with an order not based on the node names?

Sorting is used to provide a predictable, "kubeadm friendly" and consistent task exection order that can automatically adapt to any requested list of actions and to any cluster topology supported by kind. 

Such order is not based on the node name only, but on a combination of 4 attributes considered with a well known priority: node provisioning order, node name, action order, task order (I improved comments to make this more clear).

By combining above parameters you get a task execution order consistent with all the major kubeadm workflows (init, join control-plane, join) or (upgrade apply, upgrade node control-plane, upgrade node).

See discussion below with @ereslibre about implementation details, but in any case the ordering is 100% deterministic and it is already covered by unit tests signal.

As noted in `NewExecutionPlan` comments, I already know that this won't be enough for complex CI workflows, but this is a good step in that direction and this is why I'm putting order into this PR.